core/src/main/scala/scala/pickling/generator/sourcegen.scala

```diff
@@ -384,63 +384,114 @@ private[pickling]  trait SourceGenerator extends Macro with tags.FastTypeTagMacr
   }
 
   val picklingPath = q"_root_.scala.pickling"
+  val picklersRegistry = q"$picklingPath.internal.currentRuntime.picklers"
+
+  def lazyInit(lookup: c.Tree, key: c.Tree, tpe: c.Tree,
+               notInitialized: c.TermName): c.Tree = {
+    q"""
+      if($picklersRegistry.isLookupEnabled) {
+        $lookup($key) match {
+          case Some(p) => p.asInstanceOf[$tpe]
+          case None => $notInitialized: $tpe
+        }
+      } else $notInitialized: $tpe
+    """
+  }
+
+  def picklerUnpicklerLazyInit(lookup1: c.Tree, lookup2: c.Tree, key: c.Tree,
+                               tpe: c.Tree, notInitialized: c.TermName) = {
+    q"""
+      if($picklersRegistry.isLookupEnabled) {
+        ($lookup1($key), $lookup2($key)) match {
+          case (Some(p), Some(u)) if p == u => p.asInstanceOf[$tpe]
```

Comment:
maybe document that `tpe ` is the type of the *pickler* not the type being pickled.